src/hotspot/share/utilities/resourceHash.hpp

```diff
@@ -193,23 +201,40 @@ class ResourceHashtableBase : public STORAGE {
     return false;
   }
 
-  // ITER contains bool do_entry(K const&, V const&), which will be
-  // called for each entry in the table.  If do_entry() returns false,
-  // the iteration is cancelled.
-  template<class ITER>
-  void iterate(ITER* iter) const {
-    Node* const* bucket = table();
-    const unsigned sz = table_size();
-    while (bucket < bucket_at(sz)) {
-      Node* node = *bucket;
+ private:
+  template<class TABLE, class NODE, class VALUE, class ITER>
+  static void iterate_impl(TABLE* t, ITER* iter) {
+    NODE** bucket = t->table();
+    const unsigned sz = t->table_size();
+    while (bucket < t->bucket_at(sz)) {
+      NODE* node = *bucket;
       while (node != NULL) {
-        bool cont = iter->do_entry(node->_key, node->_value);
+        bool cont = iter->do_entry(const_cast<K const&>(node->_key), const_cast<VALUE>(node->_value));
```

Comment:
I don't grok the casts, why are they needed? Node is already const type here, no?